### PR TITLE
WiP: shimmed fat filter chain powered virtual outbound listener 

### PR DESCRIPTION
**Pull request by lambdai**

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/listener_builder.go [hunk 1]
```diff
@@ -15,6 +15,7 @@
 package v1alpha3
 
 import (
+	"istio.io/istio/pilot/pkg/serviceregistry"
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/listener_builder.go:18]: nit: move down

File: pilot/pkg/networking/core/v1alpha3/listener_builder.go [hunk 2]
```diff
@@ -65,6 +66,91 @@ type ListenerBuilder struct {
 	httpProxyListener       *xdsapi.Listener
 	virtualOutboundListener *xdsapi.Listener
 	virtualInboundListener  *xdsapi.Listener
+	voutBuilder             VirtualOutboundListenerInterface
+}
+
+type VirtualOutboundListenerInterface interface {
+	// Insert an outbound service into the virtual outbound listener.
+	// This function should resolve the conflict if a service is inserted repeatedly.
+	insertService(configgen *ConfigGeneratorImpl, node *model.Proxy, listenerOpts buildListenerOpts,
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/listener_builder.go:75]: I think it is: insert a listener for the specific service

File: pilot/pkg/networking/core/v1alpha3/listener_builder.go [hunk 3]
```diff
@@ -65,6 +66,91 @@ type ListenerBuilder struct {
...
+		pluginParams *plugin.InputParams,
+		virtualServices []model.Config, actualWildcard string)
+	// Invoked when build the catch rest egress listener
+	prepareDefaultEgressListener()
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/listener_builder.go:79]: confusing from the comment

**Comment by lambdai** [comment]
@hzxuzhonghu Thank you for reviewing this immature PR!

> Is this the complete one, with one listener generated? From what I see, it doesn't.

It's not complete. The interface is aimed to build the virtualInboundListener by aggregation the services.
This shimmed implementation is reusing the existing code which build the per bind_to_port=false listener including 0.0.0.0:<port> for http services and vip:<port> for tcp services.

Shim has flaws such as:
1. people will soon forget why we build per small listener and merge
2. Merging listener with listener is losing the original service information, 
3. tech debt such as the shape that 0.0.0.0:<port> actually contains http listener, and service entry such as "192.192.192.192/24", see external-svc-mongocluster in https://istio.io/docs/reference/config/networking/service-entry/
4. hard to generate the listener filter predicate

However, shim could be a control sample to prove that future virtual outbound builder is correct **when there is not conflict service/vs**. 
Even for the shim version, two extra major component is needed
1. generate the listener filter predicate, to disable sniff on server first protocol
2. develop a filter chain conflict detector in pilot so that we can give up per listener when filter chain conflict occurs. The conflict here I am referring to the conflict introduced when merging listener, such as an SE define the same VIP(or address prefix) used by another k8s service.

**Pull request closed without merging**